pkg/builder/merged_entity_meta_data_builder.go

```diff
@@ -178,124 +159,190 @@ func newPropertyBuilder(propertyName string) *propertyBuilder {
+}
+
+func newCommoditySoldMetadataBuilder(commType proto.CommodityDTO_CommodityType, ignoreIfPresent bool) *commoditySoldMetadataBuilder {
+	return &commoditySoldMetadataBuilder{
+		commodityType:   commType,
+		ignoreIfPresent: ignoreIfPresent,
+	}
+}
+
+func (builder *commoditySoldMetadataBuilder) addField(fieldName string, fieldPaths []string) *commoditySoldMetadataBuilder {
+	if fieldName == "" {
+		return builder
+	}
+	builder.fieldBuilders = append(builder.fieldBuilders,
+		newFieldBuilder(fieldName, fieldPaths))
+	return builder
+}
+
+func (builder *commoditySoldMetadataBuilder) build() *proto.MergedEntityMetadata_CommoditySoldMetadata {
+	soldMetadata := &proto.MergedEntityMetadata_CommoditySoldMetadata{
+		CommodityType:   &builder.commodityType,
+		IgnoreIfPresent: &builder.ignoreIfPresent,
+	}
+	for _, fieldBuilder := range builder.fieldBuilders {
+		soldMetadata.PatchedFields = append(soldMetadata.PatchedFields, fieldBuilder.build())
+	}
+	return soldMetadata
+}
 
+// MergedEntityMetadataBuilder is used to create an MergedEntityMetadata object.
+// MergedEntityMetadata is a message in the TemplateDTO of the supply chain.  It provides data that
+// defines the stitching behavior of entities discovered by a probe.  There should be a
+// MergedEntityMetadata entry for each entity type in the probe that is reported as origin "proxy".
+// The MergedEntityMetadata is created for stitching in XL server. It combines information that was previously
+// contained in various places used for stitching in Classic OpsManager server (e.g. external entity link, replacement
+// entity metadata, and etc.) and stores it in one place.
 type MergedEntityMetadataBuilder struct {
 	// MergedEntityMetadata - the main protobuf structure to return
 	metadata *proto.MergedEntityMetadata
-
-	// MergedEntityMetadata consists of  MatchingMetadata for proxy and external entity
+	// Deprecated: commoditiesSold exits for backward compatibility and should not be used any more
+	commoditiesSold []proto.CommodityDTO_CommodityType
+	// MergedEntityMetadata consists of MatchingMetadata for proxy (internal) and external entity
 	*matchingMetadataBuilder
-	propertyBuilders       []*propertyBuilder
-	fieldBuilders          []*fieldBuilder
-	commBoughtMetadataList []*perProviderCommodityBoughtMetadata // for different providers
+	keepStandAlone              bool
+	propertyBuilders            []*propertyBuilder
+	fieldBuilders               []*fieldBuilder
+	commBoughtMetadataList      []*commodityBoughtMetadataBuilder // for different providers
+	commoditiesSoldMetadataList []*commoditySoldMetadataBuilder
 }
 
+// NewMergedEntityMetadataBuilder initializes a MergedEntityMetadataBuilder object
 func NewMergedEntityMetadataBuilder() *MergedEntityMetadataBuilder {
 	builder := &MergedEntityMetadataBuilder{
 		metadata:                &proto.MergedEntityMetadata{},
 		matchingMetadataBuilder: newMatchingMetadataBuilder(),
+		keepStandAlone:          true,
```

Comment:
Default behavior of this field should be false. Which means after merging the proxy entity will be removed.

Reply:
The default value is true in the protocol. From the `SupplyChain.proto`:
```
message MergedEntityMetadata {
    // Should this entity type reported by the probe be kept in the topology if no stitching match
    // is found?  If yes, keepStandalone should be true.  If no, it should be false.
    optional bool keepStandalone = 1 [default = true];
    ...
    ...
}
```
The field seems to indicate whether to keep the proxy entity in the topology if there is no match found during stitching. After a successful match, the proxy entities will always be dropped.

Reply:
Ok to keep the default (true) consistent with the protobuf message.